# Programming Assignment - Analysing Product Sentiment

In [1]:
import pandas as pd
import turicreate as tc

In [2]:
products = tc.SFrame('amazon_baby.sframe')
products.head()

Columns:
	name	str
	review	str
	rating	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[10 rows x 3 columns]

In [3]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

**1a. Use .apply() to build new features with the counts for each of the selected words**

In [4]:
# create the word count feature
products['word_count'] = tc.text_analytics.count_words(products['review'])

tc.config.set_runtime_config('TURI_DEFAULT_NUM_PYLAMBDA_WORKERS', 32)
# iterate overall words
for word in selected_words:
    # extract count (value) if word (key) in word_count dictionary, else return 0
    products[word] = products.apply(lambda x: x['word_count'][word] if word in x['word_count'] else 0)
#     products[word] = products.apply(lambda x: x['word_count'].get(word) if word in x['word_count'] else 0)
products.head(5)

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	int
	great	float
	fantastic	float
	amazing	int
	love	float
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'handles': 1.0, 'strippin... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'recommend': 1.0, 'disapp... |    0    |  0.0  |    0.0    |    0    | 1.0  |
| {'quilt': 1.0, 'the': 1.0,... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'tool': 1.0, 'clever': 1.... |    0    |  0.0  |    0.0    |    0    | 2.0  |
| {'rock': 1.0, 'many': 1.0,... |    0    |  1.0  |    0.0    |    0    | 1.0  |
| {'thumb': 1.0, 'or': 1.0, ... |    0    |  1.0  |    0.0    |    0    | 0.0  |
| {'for': 1.0, 'barnes': 1.0... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'right': 1.0, 'because': ... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'like': 1.0, 'and': 1.0, ... |    0    |  0.0  |    1.0    |    0    | 0.0  |
| {'in': 1.0, 'pages': 1.0, ... |    0    |  0.0  |    0.0    |    0    | 0.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+
| horrible | bad | terrible | awful | wow | hate |
+----------+-----+----------+-------+-----+------+
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
+----------+-----+----------+-------+-----+------+
[183531 rows x 15 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

**1b. Use .sum() to identify the most-used and least-used words**

In [12]:
print("word counts")
for word in selected_words:
    count = products[word].sum()
    print('{word}: {count}'.format(word=word, count=count))

word counts
awesome: 4075
great: 59536.0
fantastic: 1765.0
amazing: 2726
love: 43867.0
horrible: 1245
bad: 4950
terrible: 1282
awful: 753
wow: 461
hate: 1285


The most-used word in the list is: 'great' (closely followed by 'love')
The least-used word in the list is: 'wow' (followed by 'awful')

**2. Create a sentiment analysis model using only selected_words as features**

In [13]:
# ignore all 3* reviews
products = products[products['rating'] != 3]

# positive sentiment = 4-star or 5-star reviews
products['sentiment'] = products['rating'] >= 4
products.head(2)

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	int
	great	float
	fantastic	float
	amazing	int
	love	float
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'recommend': 1.0, 'disapp... |    0    |  0.0  |    0.0    |    0    | 1.0  |
| {'quilt': 1.0, 'the': 1.0,... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'tool': 1.0, 'clever': 1.... |    0    |  0.0  |    0.0    |    0    | 2.0  |
| {'rock': 1.0, 'many': 1.0,... |    0    |  1.0  |    0.0    |    0    | 1.0  |
| {'thumb': 1.0, 'or': 1.0, ... |    0    |  1.0  |    0.0    |    0    | 0.0  |
| {'for': 1.0, 'barnes': 1.0... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'right': 1.0, 'because': ... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'like': 1.0, 'and': 1.0, ... |    0    |  0.0  |    1.0    |    0    | 0.0  |
| {'in': 1.0, 'pages': 1.0, ... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'tracker': 1.0, 'now': 1.... |    0    |  0.0  |    0.0    |    0    | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
[166752 rows x 16 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [23]:
train_data, test_data = products.random_split(0.8, seed=0)

In [24]:
selected_words_model = tc.logistic_classifier.create(train_data, target='sentiment', features=selected_words,
                                                     validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients      : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.265023     | 0.847401          | 0.845874            |

| 2         | 3        | 0.469599     | 0.847514          | 0.846085            |

| 3         | 4        | 0.767944     | 0.847626          | 0.846115            |

| 4         | 5        | 0.969383     | 0.847708          | 0.846385            |

| 5         | 6        | 1.265823     | 0.847708          | 0.846385            |

| 6         | 7        | 1.469487     | 0.847708          | 0.846385            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [30]:
# view the coefficients
model_coefficients = selected_words_model.coefficients.sort(key_column_names='value')
model_coefficients.print_rows(num_rows=12)

+-------------+-------+-------+-----------------------+----------------------+
|     name    | index | class |         value         |        stderr        |
+-------------+-------+-------+-----------------------+----------------------+
|   horrible  |  None |   1   |  -2.2513352367590986  | 0.08020249388788438  |
|   terrible  |  None |   1   |  -2.2236614360851323  | 0.07731736203785755  |
|    awful    |  None |   1   |   -2.052908204031357  | 0.10099735435259254  |
|     hate    |  None |   1   |  -1.3484407222463173  | 0.07715698604297334  |
|     bad     |  None |   1   |  -0.9914778800650625  | 0.03848428664699068  |
|     wow     |  None |   1   | -0.009538236067684629 | 0.16046411224711657  |
|    great    |  None |   1   |   0.8630655001196574  | 0.01895505244437721  |
|  fantastic  |  None |   1   |   0.8858047568814237  | 0.11167591293399677  |
|   amazing   |  None |   1   |   1.1000933113660223  | 0.09954776260465992  |
|   awesome   |  None |   1   |   1.1335346660341379

The coefficient with the largest positive weight is: 'love'
The coefficient with the largest negative weight is: 'awesome'

These results partly make sense. Nearly all the coefficients are all in the right direction, but I would have thought that 'great' would have a larger coefficient given that it was used the most out of all the words. Similarly, 'bad' was used the most out of all the negative words, and yet it has a very low weight.

**3. Compare the accuracy of different sentiment analysis models**

In [31]:
sentiment_model = tc.logistic_classifier.create(train_data, target='sentiment', features=['word_count'],
                                                validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 57356

Number of coefficients      : 57357

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 4        | 0.250000  | 2.143929     | 0.841421          | 0.840019            |

| 1         | 9        | 3.250000  | 4.643570     | 0.931359          | 0.911362            |

| 2         | 11       | 2.669978  | 5.841241     | 0.938650          | 0.916466            |

| 3         | 12       | 2.669978  | 6.633002     | 0.927417          | 0.901814            |

| 4         | 14       | 1.326072  | 7.736259     | 0.945200          | 0.918088            |

| 9         | 20       | 1.326072  | 11.936829    | 0.977654          | 0.917698            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [32]:
acc_score1 = selected_words_model.evaluate(test_data)['accuracy']
acc_score1

0.8463848186404036

In [33]:
acc_score2 = sentiment_model.evaluate(test_data)['accuracy']
acc_score2

0.9176975738650012

In [41]:
baseline_acc = products['sentiment'].value_counts()[0]['count'] / products.shape[0]
baseline_acc

0.8411233448474381

Both the full and selected_words models beat the baseline/majority class classifier, but the selected_words model beats it only by a slim margin.

**4. Interpret the difference in performance**

In [44]:
# filter the dataset to only include the product 'Baby Trend Diaper Champ'
diaper_champ_reviews = products[products['name'] == 'Baby Trend Diaper Champ']
diaper_champ_reviews.head(2)

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	int
	great	float
	fantastic	float
	amazing	int
	love	float
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int

Rows: 2

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'convenient': 1.0, 'more'... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'system': 1.0, 'try': 1.0... |    0    |  0.0  |    0.0    |    0    | 0.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     0     |
+----------+-----+----------+-------+-----+------+-----------+
[2 rows x 16 columns]

In [52]:
# predict the sentiment using the full model
diaper_champ_reviews['sentiment_model_predictions'] = sentiment_model.predict(diaper_champ_reviews, output_type='probability')
# predict the sentiment using the selected_words model
diaper_champ_reviews['selected_words_predictions'] = selected_words_model.predict(diaper_champ_reviews,
                                                                                  output_type='probability')
# order the results by the full model results
diaper_champ_reviews = diaper_champ_reviews.sort(key_column_names='sentiment_model_predictions', ascending=False)
diaper_champ_reviews.head(5)

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	int
	great	float
	fantastic	float
	amazing	int
	love	float
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	sentiment_model_predictions	float
	selected_words_predictions	float

Rows: 5

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I read a review below that... |  4.0   |
| Baby Trend Diaper Champ | I have never written a rev... |  5.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | Baby Luke can turn a clean... |  5.0   |
| Baby Trend Diaper Champ | Diaper Champ or Diaper Gen... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'key': 1.0, 'have': 1.0, ... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'priceless': 1.0, 'knows'... |    0    |  0.0  |    0.0    |    0    | 1.0  |
| {'price': 1.0, 'suggestion... |    0    |  0.0  |    0.0    |    0    | 0.0  |
| {'around': 1.0, 'any': 1.0... |    0    |  1.0  |    0.0    |    0    | 0.0  |
| {'either': 1.0, 'be': 1.0,... |    0    |  1.0  |    0.0    |    0    | 0.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
|    0     |  0  |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+-----------------------------+----------------------------+
| sentiment_model_predictions | selected_words_predictions |
+-----------------------------+----------------------------+
|      0.9999999999895941     |     0.7919288370624469     |
|      0.9999999999868131     |     0.9367819244799372     |
|      0.9999999999465673     |     0.7919288370624469     |
|      0.9999999999302822     |     0.9002186948093646     |
|      0.9999999999174133     |     0.9002186948093646     |
+-----------------------------+----------------------------+
[5 rows x 18 columns]

The predicted sentiment for the most positive review is: 0.9999999999895941.

The predicted sentiment for that same review according to the selected words model is: 0.7919288370624469

In [54]:
diaper_champ_reviews['review'][0]

"I read a review below that can explain exactly what we experienced. We've had it for 16 months and it has worked wonderful for us. No smells, change it out once a week, easy to clean. Then a diaper snagged this foam material in the head part, so I pulled the rest of the foam out. Big mistake!!! Now it can no loner retain the stinkiness and we're looking for a replacement. Be careful of overloading and never take out that foam piece that is cushioned between pieces. I have figured out that it is key to keeping the stink out."

In [63]:
selected_words

['awesome',
 'great',
 'fantastic',
 'amazing',
 'love',
 'horrible',
 'bad',
 'terrible',
 'awful',
 'wow',
 'hate']

In [62]:
diaper_champ_reviews['word_count'][0]

{'key': 1.0,
 'have': 1.0,
 'pieces': 1.0,
 'between': 1.0,
 'cushioned': 1.0,
 'piece': 1.0,
 'take': 1.0,
 'overloading': 1.0,
 'be': 1.0,
 'looking': 1.0,
 're': 1.0,
 'stinkiness': 1.0,
 'retain': 1.0,
 'now': 1.0,
 'wonderful': 1.0,
 'worked': 1.0,
 '16': 1.0,
 'and': 3.0,
 'months': 1.0,
 've': 1.0,
 'in': 1.0,
 'us': 1.0,
 'i': 3.0,
 'experienced': 1.0,
 'read': 1.0,
 'easy': 1.0,
 'for': 3.0,
 'to': 2.0,
 'has': 1.0,
 'review': 1.0,
 'keeping': 1.0,
 'replacement': 1.0,
 'out': 5.0,
 'loner': 1.0,
 'clean': 1.0,
 'mistake': 1.0,
 'big': 1.0,
 'pulled': 1.0,
 'it': 5.0,
 'this': 1.0,
 'is': 2.0,
 'explain': 1.0,
 'material': 1.0,
 'exactly': 1.0,
 'a': 4.0,
 'we': 3.0,
 'that': 4.0,
 'had': 1.0,
 'what': 1.0,
 'part': 1.0,
 'no': 2.0,
 'smells': 1.0,
 'can': 2.0,
 'change': 1.0,
 'figured': 1.0,
 'week': 1.0,
 'then': 1.0,
 'snagged': 1.0,
 'diaper': 1.0,
 'careful': 1.0,
 'the': 5.0,
 'never': 1.0,
 'foam': 3.0,
 'head': 1.0,
 'so': 1.0,
 'below': 1.0,
 'rest': 1.0,
 'stink': 1

The value of the predicted sentiment for the most positive review using the full model is much more positive than the value predicted using the selected_words model because none of the words in selected_words appeared in the text of this review.